coderd/database/dbtestutil/broker.go

```diff
@@ -0,0 +1,176 @@
+	if openOptions.LogDSN {
+		_, _ = fmt.Fprintf(os.Stderr, "Connect to the database for %s using: psql '%s'\n", t.Name(), testDBParams.DSN())
+	}
+	t.Cleanup(b.clean(t, dbName))
+	return testDBParams, nil
+}
+
+func (b *Broker) clean(t TBSubset, dbName string) func() {
+	return func() {
+		_, err := b.coderTestingDB.Exec("DROP DATABASE " + dbName + ";")
+		if err != nil {
+			t.Logf("failed to clean up database %q: %s\n", dbName, err.Error())
+			return
+		}
+		_, err = b.coderTestingDB.Exec("UPDATE test_databases SET dropped_at = CURRENT_TIMESTAMP WHERE name = $1", dbName)
+		if err != nil {
+			t.Logf("failed to mark test database '%s' dropped: %s\n", dbName, err.Error())
+		}
+	}
+}
+
+func (b *Broker) init(t TBSubset) error {
+	b.Lock()
+	defer b.Unlock()
+	b.refCount++
+	t.Cleanup(b.decRef)
+	if b.coderTestingDB != nil {
+		// already initialized
+		return nil
+	}
+
+	connectionParamsInitOnce.Do(func() {
+		errDefaultConnectionParamsInit = initDefaultConnection(t)
+	})
+	if errDefaultConnectionParamsInit != nil {
+		return xerrors.Errorf("init default connection params: %w", errDefaultConnectionParamsInit)
+	}
+	coderTestingParams := defaultConnectionParams
+	coderTestingParams.DBName = CoderTestingDBName
+	coderTestingDB, err := sql.Open("postgres", coderTestingParams.DSN())
+	if err != nil {
+		return xerrors.Errorf("open postgres connection: %w", err)
+	}
+
+	// creating the db can succeed even if the database doesn't exist. Ping it to find out.
+	err = coderTestingDB.Ping()
+	var pqErr *pq.Error
+	if xerrors.As(err, &pqErr) && pqErr.Code == "3D000" {
+		// database does not exist.
+		if closeErr := coderTestingDB.Close(); closeErr != nil {
+			return xerrors.Errorf("close postgres connection: %w", closeErr)
+		}
+		err = createCoderTestingDB(t)
+		if err != nil {
+			return xerrors.Errorf("create coder testing db: %w", err)
+		}
+		coderTestingDB, err = sql.Open("postgres", coderTestingParams.DSN())
+		if err != nil {
+			return xerrors.Errorf("open postgres connection: %w", err)
+		}
+	} else if err != nil {
+		_ = coderTestingDB.Close()
+		return xerrors.Errorf("ping '%s' database: %w", CoderTestingDBName, err)
+	}
+	b.coderTestingDB = coderTestingDB
+	b.uuid = uuid.New()
+	return nil
+}
+
+func createCoderTestingDB(t TBSubset) error {
+	db, err := sql.Open("postgres", defaultConnectionParams.DSN())
+	if err != nil {
+		return xerrors.Errorf("open postgres connection: %w", err)
+	}
+	defer func() {
+		_ = db.Close()
+	}()
+	err = createAndInitDatabase(t, defaultConnectionParams, db, CoderTestingDBName, func(testDB *sql.DB) error {
+		_, err := testDB.Exec(coderTestingSQLInit)
+		return err
+	})
+	if err != nil {
+		return xerrors.Errorf("create coder testing db: %w", err)
+	}
+	return nil
+}
+
+func (b *Broker) decRef() {
+	b.Lock()
+	defer b.Unlock()
+	b.refCount--
+	if b.refCount == 0 {
+		// ensures we don't leave go routines around for GoLeak to find.
+		_ = b.coderTestingDB.Close()
+		b.coderTestingDB = nil
+	}
```

Comment:
How much connection churn does this create, and is it a problem? I'd be worried that this would create hundreds of new testdb connections in packages with many test databases

Reply:
Without this refcounted design, we are stuck creating a new connection for each test case. This is designed to _reduce_ connection churn by allowing the connections to be reused.

Note that the previous implementation created a new database for each call to `dbtestutil.Open()`. Since we run tests in parallel, I expect connection reuse to be relatively high with this design.